### DISCUSSION TITLE
Writing tiles to numpy in parallel

### DISCUSSION THREAD

**Discussion by edwinytgoh** [pydata/xarray#7473] [General]
I am trying to improve the performance of my script that takes an xarray dataset containing a variable with shape (~2e6, 256, 256) that corresponds to the coordinates `index`, `i_local`, and `j_local`, respectively. My goal is to cast the entire variable to float16 and write each `index` to individual 256x256 PNG files.

This dataset is a subset of a 5PB ocean simulation (LLC4320) that is lazily loaded using xarray, so I'm essentially streaming from one location on disk into numpy files in another location on disk. However, my current implementation is very slow, taking ~8 seconds per tile. I have looked into using map_blocks and GroupBy.apply, but those seem to require helper functions that return arrays, whereas my `save_to_numpy` function doesn't return anything.

Is there any way to parallelize this operation using xarray and dask? Are there any best practices for this type of operation? Any help would be greatly appreciated.

---
Just in case, here's my `save_to_numpy ` function:
```
import io
import tarfile
from tqdm.auto import tqdm

def save_to_numpy(ds, out_folder):
    # convert to single precision float16 to save space
    ds = ds.astype("float16").chunk({"index": 2048})
    
    # write numpy tiles to tar file using io.BytesIO()    
    tar_path = os.path.join(out_folder, "output.tar")
    with tarfile.open(tar_path, "a") as tar:
        for idx in tqdm(ds.index.values):  # each iteration takes 8 seconds!
                tile = get_numpy_tile(ds["Theta"].isel(index=idx))
                filename = f"{idx}.npy"
                with io.BytesIO() as buf:
                    np.save(buf, tile)
                    buf.seek(0)
                    info = tarfile.TarInfo(filename)
                    info.size = len(buf.getvalue())
                    tar.addfile(tarinfo=info, fileobj=buf)        

def get_numpy_tile(tile):
    tile = tile.drop_vars(["index", "time", "tile_num"])
    tile = tile.squeeze()
    tile = tile.transpose("j_local", "i_local")
    tile = tile.load()
    tile = tile.values
    tile = tile.astype(np.float16)
    return tile
```

**Comment by TomNicholas** [comment]
You could always just use dask.delayed [like we did here](https://gist.github.com/TomNicholas/ee5c673f8772beeecd3f58f9b040943f). Basically create a function which when called pulls out one chunk of your array and writes just that chunk to a file. Then wrap that function in dask.delayed and I think dask should be able to parallelize that.

**Reply by edwinytgoh** [reply]
Thanks for the reply! I wrote a new function called `save_to_numpy_parallel` and decorated my `save_tile` function with `@dask.delayed`. However, I encountered the following error upon execution: 
```
  File "src/generate_llc_tiles.py", line 111, in save_to_numpy_parallel
    results = dask.compute(*delayed_ops, num_workers=20)
  File "/python3.8/site-packages/dask/base.py", line 573, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/python3.8/site-packages/dask/multiprocessing.py", line 220, in get
    result = get_async(
  File "/python3.8/site-packages/dask/local.py", line 494, in get_async
    fire_tasks(chunksize)
  File "/python3.8/site-packages/dask/local.py", line 476, in fire_tasks
    dumps((dsk[key], data)),
  File "/python3.8/site-packages/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/python3.8/site-packages/cloudpickle/cloudpickle_fast.py", line 632, in dump
    return Pickler.dump(self, obj)
TypeError: cannot pickle '_io.BufferedWriter' object
```
Do you have any suggestions on how to address this error? 

My modified `save_to_numpy_parallel` function is as follows:
```
import dask
dask.config.set(scheduler='processes')
dask.config.set({"distributed.worker.use-file-locking": False})    

def save_to_numpy_parallel(ds, out_folder):
    # convert to single precision float16 to save space
    ds = ds.astype("float16").chunk({"index": 2048})
    # write numpy tiles to tar file using io.BytesIO()    
    tar_path = os.path.join(out_folder, "output.tar")
    with tarfile.open(tar_path, "a") as tar:
        delayed_ops = []  # for use with dask.compute()
        for idx in ds.index.values:
            delayed_ops.append(save_tile(ds, idx, tar))
        # Compute the delayed operations
        results = dask.compute(*delayed_ops, num_workers=20)

@dask.delayed
def save_tile(ds, idx, tar):
    tile = get_numpy_tile(ds["Theta"].isel(index=idx))
    with io.BytesIO() as buf:
        np.save(buf, tile)
        buf.seek(0)
        info = tarfile.TarInfo(f"{idx}.npy")
        info.size = len(buf.getvalue())
        tar.addfile(tarinfo=info, fileobj=buf)
···

**Reply by TomNicholas** [reply]
I'm afraid I'm not sure :confused: There are [some results for that error](https://github.com/marcelm/cutadapt/issues/469) online though, I would start there.

**Reply by TomNicholas** [reply]
Worth also asking on the [pangeo discourse](https://discourse.pangeo.io/) too, with an explanation of why you're trying to achieve what you're doing.

**Reply by dcherian** [reply]
Seems like you can't use cloudpickle with that io thing,. Dask needs this to communicate between workers. 

I would write out a simple dask.array version of the problem and ask on the dask tracker.

**Reply by edwinytgoh** [reply]
> Worth also asking on the [pangeo discourse](https://discourse.pangeo.io/) too, with an explanation of why you're trying to achieve what you're doing.

Thanks for the suggestion, I'll check out the pangeo discourse as well. I actually just responded to @rabernat below about what we're trying to use the dataset for, and why we're thinking about saving each tile as a separate .npy file. 

I also attempted to remove the tar writing portion of my `save_to_numpy_parallel` function so that each dask delayed op directly writes to a single .npy file, but it resulted in an out-of-memory issue even with 20 workers. I'll try to reduce the number of workers and report back.

**Comment by rabernat** [comment]
Why not just use Zarr? That's basically what Zarr does: split arrays into a bunch of chunks.

**Reply by edwinytgoh** [reply]
We initially tried using Zarr in an ML pipeline (PyTorch) but found that each batch took too long (~1 minute) to load. This is why we considered emulating the good I/O performance of large computer vision benchmarks (e.g., ImageNet) by storing images as separate JPG files and letting PyTorch handle the [multi-process data loading](https://pytorch.org/docs/stable/data.html#multi-process-data-loading). We'd really appreciate any suggestions or best practices for integrating xarray/Zarr with PyTorch's data loader API! 

---
This is our initial Zarr dataset:
![image](https://user-images.githubusercontent.com/37746563/215205553-5591a17e-5d58-4363-b9cc-a208524f4180.png)

I have a small, PyTorch-independent benchmark using the following snippet:
```
import numpy as np
import xarray as xr
ds = xr.open_dataset("/scratch/LLC4320/SST/2012.zarr").chunk({"index": 2048})  # /scratch is a Lustre filesystem
sst = ds["Theta"]
# profile the time taken to load a "batch" of 256 random tiles (~32MB per batch)
get_batch = lambda sst: sst[..., np.random.randint(0, 300000, 256)].values
get_batch(sst)  # warm up run
%timeit -n 4 -r 10 get_batch(sst)  #  single-worker analogue to 4 epochs, 10 batches per epoch
```
This gives: `32.1 s ± 3.4 s per loop (mean ± std. dev. of 10 runs, 4 loops each)`, which is too slow for ML purposes. Our goal is to have <500ms per batch.

I also thought it could be an issue with the row-major vs. column-major order  ((`j_local`, `i_local`, `index`) vs. (`index`, `j_local`, `i_local`)), but reading from a transposed Zarr store didn't seem to make much of a difference.

**Reply by rabernat** [reply]
This is a really interesting use case. I'm going to spend some time looking into it. But I'd like to get a bit more info from you.

```python
ds = xr.open_dataset("/scratch/LLC4320/SST/2012.zarr").chunk({"index": 2048})  # /scratch is a Lustre filesystem
```

This makes it look like you're rechunking the data on the fly. Why? What is the internal chunking of the Zarr array?

The key question is _what is the actual Zarr chunking on disk_.

```python
get_batch = lambda sst: sst[..., np.random.randint(0, 300000, 256)].values
```

This is likely the core problem. You're trying to shuffle the data. Depending on how the data are chunked, you are most likely having to load every chunk in every batch. The solution is to use smaller chunks on disk (which is effectively what you are doing by exporing single images). But without information about your chunks on disk, it's hard to know for sure. 

---

@maxrjones and other folks have been working on the [xbatcher](https://github.com/xarray-contrib/xbatcher) project, which is aimed pretty squarely at this use case. I'm wondering if max has some ideas.

**Reply by edwinytgoh** [reply]
The `.chunk({"index": 2048})` call makes use of all cores when running `get_batch()`, which otherwise only uses 1 core/thread. However, it doesn't seem to meaningfully affect timings.
<img src="https://user-images.githubusercontent.com/37746563/215641783-2a9d6ee3-8c5b-4c47-907f-3a9fbcf5dd0e.png" width="550">

In terms of the chunk size on disk, here's a sample output from `lfs getstripe 2012.zarr`:
```
2012.zarr/Theta
stripe_count:  1 stripe_size:   1048576 pattern:       raid0 stripe_offset: -1

./0.0.306
lmm_stripe_count:  1
lmm_stripe_size:   1048576
lmm_pattern:       raid0
lmm_layout_gen:    0
lmm_stripe_offset: 1
        obdidx           objid           objid           group
             1         5859423       0x59685f                0

./0.0.204
lmm_stripe_count:  1
lmm_stripe_size:   1048576
lmm_pattern:       raid0
lmm_layout_gen:    0
lmm_stripe_offset: 5
        obdidx           objid           objid           group
             5         7938280       0x7920e8                0

./0.0.178
lmm_stripe_count:  1
lmm_stripe_size:   1048576
lmm_pattern:       raid0
lmm_layout_gen:    0
lmm_stripe_offset: 6
        obdidx           objid           objid           group
             6         7977223       0x79b907                0

```

Thanks for the suggestion to use `xbatch`! I will give that a shot after I figure out the individual numpy tiles. In the meantime, I'll also try running a benchmark with `get_batch = lambda sst: sst[..., np.arange(0, 256)].values` to verify whether the issue really lies with the random sampling.

**Reply by edwinytgoh** [reply]
```
import numpy as np
import xarray as xr
ds = xr.open_dataset("/scratch/LLC4320/SST/2012.zarr")
sst = ds["Theta"]
# profile the time taken to load the "batch" of first 256 tiles (~32MB per batch)
get_batch = lambda sst: sst[..., np.arange(0, 256)].values
get_batch(sst)  # warm up run
%timeit -n 4 -r 10 get_batch(sst)  #  single-worker analogue to 4 epochs, 10 batches per epoch
```
Gives: _252 ms ± 5.09 ms per loop (mean ± std. dev. of 10 runs, 4 loops each)_

### Manual chunking with `.chunk()`
`ds = xr.open_dataset("/scratch/LLC4320/SST/2012.zarr").chunk({"index": 2048})`:  _495 ms ± 8.06 ms per loop_

### Different indexing variations:
- 2048 "batch", `np.arange(0, 2048)`: _882 ms ± 8.22 ms per loop_
- 2048 "batch" but reading across chunks, `np.arange(1024, 3072)`: _1.13 s ± 12.5 ms per loop_
- 4096 "batch", `np.arange(0, 4096)`:  _1.8 s ± 13.4 ms per loop_

**Reply by rabernat** [reply]
@edwinytgoh, you're still not understanding my question about the chunks on disk. I apologize for not being clear.

Please run the following code and share the output

```python
import zarr
group = zarr.open("/scratch/LLC4320/SST/2012.zarr/")
print(group["Theta"].info)
```

**Reply by edwinytgoh** [reply]
Ah, I apologize for misunderstanding your question - I was under the impression that you were talking about actual hardware-level chunking across the Lustre filesystem. Here's the output from `group["Theta"].info`:

```
Name               : /Theta
Type               : zarr.core.Array
Data type          : float16
Shape              : (256, 256, 897920)
Chunk shape        : (256, 256, 2048)
Order              : C
Read-only          : False
Compressor         : Blosc(cname='lz4', clevel=5, shuffle=SHUFFLE, blocksize=0)
Store type         : zarr.storage.DirectoryStore
No. bytes          : 117692170240 (109.6G)
No. bytes stored   : 73543091370 (68.5G)
Storage ratio      : 1.6
Chunks initialized : 439/439
```

**Reply by rabernat** [reply]
OK, excellent, so this confirms that you chunks on disk are indeed 2048
```
Chunk shape        : (256, 256, 2048)
```

If you really want to shuffle your data over the full `index` axis, you might want to consider writing a copy with 

```
Chunk shape        : (256, 256, 1)
```

This should be equivalent to what you're trying to do with numpy tiles: one chunk per image.

Also, experiment with not using Dask at all. Xarray by default uses lazy array access and won't load the data until you need it. Also remember that whenever you call `.chunk`, everything downstream of that has to load the entire chunk into memory.

**Comment by negin513** [comment]
You might want to consider using [xbatcher](https://github.com/xarray-contrib/xbatcher) which is developed for ML applications. It basically makes it easy to feed Xarray objects to machine learning libraries such as PyTorch or Keras. 
[Here](https://xbatcher.readthedocs.io/en/latest/demo.html) is a good example of using Zarr + Xbatcher.
